app/shared/app-data/src/commonMain/kotlin/domain/session/AniAuthConfigurator.kt

```diff
@@ -0,0 +1,355 @@
+import me.him188.ani.utils.logging.logger
+import me.him188.ani.utils.platform.Uuid
+import me.him188.ani.utils.platform.currentTimeMillis
+import kotlin.coroutines.CoroutineContext
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.days
+import kotlin.time.Duration.Companion.seconds
+
+/**
+ * Wrapper for [SessionManager] and [AniAuthClient] to handle authorization.
+ * Usually use it at UI layer (e.g. at ViewModel).
+ * 
+ * You should call [authorizeRequestCheckLoop] to start the test runner loop and make functionality.
+ */
+class AniAuthConfigurator(
+    private val sessionManager: SessionManager,
+    private val authClient: AniAuthClient,
+    private val onLaunchAuthorize: suspend (requestId: String) -> Unit,
+    private val networkMaxRetries: Long = 10,
+    private val awaitRetryInterval: Duration = 1.seconds,
+    parentCoroutineContext: CoroutineContext = Dispatchers.Default,
+) {
+    private val logger = logger<AniAuthConfigurator>()
+    private val scope = parentCoroutineContext.childScope()
+
+    private val authorizeTasker = MonoTasker(scope)
+    private val currentRequestAuthorizeId = MutableStateFlow<String?>(null)
+
+    val state: StateFlow<AuthStateNew> = currentRequestAuthorizeId
+        .transformLatest { requestId ->
+            if (requestId == null) {
+                emit(AuthStateNew.Idle)
+                logger.debug { "[AuthState] Got null request id, stopped checking." }
+                return@transformLatest
+            } else {
+                emit(AuthStateNew.AwaitingResult(requestId))
+            }
+
+            logger.debug { "[AuthState][${requestId.idStr}] Start checking session state." }
+            sessionManager.state
+                .collectLatest { sessionState ->
+                    // 如果有 token, 直接获取当前 session 的状态即可
+                    if (sessionState !is SessionStatus.NoToken)
+                        return@collectLatest collectCombinedAuthState(requestId, sessionState, null)
+
+                    // 如果是 NoToken 并且还是 REFRESH, 则直接返回 Idle
+                    if (requestId == REFRESH) {
+                        logger.debug { "[AuthState][${requestId.idStr}] No existing session." }
+                        emit(AuthStateNew.Idle)
+                        return@collectLatest
+                    }
+                    
+                    sessionManager.processingRequest
+                        .transform { processingRequest ->
+                            if (processingRequest == null) {
+                                logger.debug { "[AuthState][${requestId.idStr}] No processing request." }
+                                emit(null)
+                            } else {
+                                logger.debug { "[AuthState][${requestId.idStr}] Current processing request: $processingRequest" }
+                                emitAll(processingRequest.state)
+                            }
+                        }
```

Comment:
this is essentially:

```suggestion
                        .flatMap { processingRequest ->
                            if (processingRequest == null) {
                                logger.debug { "[AuthState][${requestId.idStr}] No processing request." }
                                flowOf(null)
                            } else {
                                logger.debug { "[AuthState][${requestId.idStr}] Current processing request: $processingRequest" }
                                processingRequest.state
                            }
                        }
```

特定的操作符比 transform 要容易理解逻辑.